shared/tinyusb/mp_usbd_cdc.c

```diff
@@ -83,40 +85,50 @@ void tud_cdc_rx_cb(uint8_t itf) {
 
 mp_uint_t mp_usbd_cdc_tx_strn(const char *str, mp_uint_t len) {
     size_t i = 0;
-    if (tud_cdc_connected()) {
-        while (i < len) {
-            uint32_t n = len - i;
-            if (n > CFG_TUD_CDC_EP_BUFSIZE) {
-                n = CFG_TUD_CDC_EP_BUFSIZE;
-            }
-            int timeout = 0;
-            // Wait with a max of USC_CDC_TIMEOUT ms
-            while (n > tud_cdc_write_available() && timeout++ < MICROPY_HW_USB_CDC_TX_TIMEOUT) {
-                mp_event_wait_ms(1);
+    while (i < len) {
+        uint32_t n = len - i;
+        if (n > CFG_TUD_CDC_EP_BUFSIZE) {
+            n = CFG_TUD_CDC_EP_BUFSIZE;
+        }
+        int timeout = 0;
+        // Wait with a max of USC_CDC_TIMEOUT ms
+        while (n > tud_cdc_write_available()
+               && timeout++ < MICROPY_HW_USB_CDC_TX_TIMEOUT
+               && tud_cdc_connected()) {
+            mp_event_wait_ms(1);
 
-                // Explicitly run the USB stack as the scheduler may be locked (eg we
-                // are in an interrupt handler), while there is data pending.
-                mp_usbd_task();
-            }
-            if (timeout >= MICROPY_HW_USB_CDC_TX_TIMEOUT) {
-                break;
-            }
-            uint32_t n2 = tud_cdc_write(str + i, n);
-            tud_cdc_write_flush();
-            i += n2;
+            // Explicitly run the USB stack as the scheduler may be locked (eg we
+            // are in an interrupt handler), while there is data pending.
+            mp_usbd_task();
+        }
+        n = MIN(n, tud_cdc_write_available());
+        if (n == 0) {
+            break;
```

Comment:
IIUC, if disconnected then after the tinyusb FIFO is full then further output is dropped until USB-CDC connects.

Could this lead to some misleading situations where USB host connects and receives:

> banner
> early program output
> [unknown number of dropped bytes here]
> contemporary output that was sent after USB host connected

Is that right? Does it seem like a problem?

The best solution I can see from this perspective would be to rotate bytes out of the overflowing FIFO as needed, but (a) this is quite slow if USB never connects and (b) not supported by current TinyUSB API.

However there is `tud_cdc_n_write_clear()` which empties the write FIFO. What do you think about doing this each time the FIFO fills up while disconnected? It means maybe some buffered bytes are lost, but the "new after old with missing middle" situation shown above can't happen.

(Note this isn't a request necessarily, I'm not fully convinced this is worth worrying about.)

Related request I do have is that it'd be good to have a short comment or so in here please, outlining the intended execution flow for this "disconnected, buffering" case as it may not be obvious to future readers why it's structured like this.

Reply:
Yeah you raise a good question. My thoughts were that keeping the banner (and maybe a little after it) is a good thing to display on connection. 
But yes it could be somewhat confusing if there's a running application with more information, the disconnect in log could seem quite random. I'm not too sure how STM handles this currently, whether it's the initial data persisted or the latest data. @dpgeorge do you have a preference here?

Reply:
In stm32 it's a rolling ring buffer, so new data is always placed into the pending CDC output buffer, and if the ring buffer is full then the oldest data is lost.  That means on connection to a USB host, the host sees a consistent history of the most recent data, possibly truncated at the beginning of that data.

Reply:
Thanks @dpgeorge, the consistent view of "latest data" does make sense really. I'll have a closer look to see if I can make it automatically overwrite, I actually think this might be possible. 

If not, the suggestion of "clear on full" might be worth doing instead.

Reply:
Hah, this is convenient: https://github.com/hathach/tinyusb/blob/1f259b3ab0f1b9587dee9c0bfb5574d70d021640/src/class/cdc/cdc_device.c#L250

So yeah if I basically remove the space in buffer check when not connected, it'll pretty much "just work". Though I'll need to confirm behavior when you try to write more than a buffer's worth in one go, whether it keeps the start or end of the data.

Reply:
That is certainly convenient!

It may actually be impossible to debug this with mpremote, given the pyserial hard-coded flush of the input buffer on open. :facepalm: But it should be visible with other serial programs, I'd expect.

Reply:
I made the change and it seemed to work fine! Even with mpremote.... I added a main.py on the samd51 board:  
`print(list(range(512)))`

Now when I start mpremote I see:
```
 $ mpremote
Connected to MicroPython at /dev/ttyACM0
Use Ctrl-] or Ctrl-x to exit this shell
 486, 487, 488, 489, 490, 491, 492, 493, 494, 495, 496, 497, 498, 499, 500, 501, 502, 503, 504, 505, 506, 507, 508, 509, 510, 511]
MicroPython v1.23.0-preview.379.g130e763226.dirty on 2024-05-25; ItsyBitsy M4 ExpreType "help()" for more information.
>>>
```

Reply:
This behaviour is perfect!